File: payment-service/src/main/java/com/ecommerce/paymentservice/global/exception/ErrorResponse.java
package com.ecommerce.paymentservice.global.exception;

import lombok.Getter;
import org.springframework.http.ResponseEntity;

@Getter
public class ErrorResponse {

    private final String code;
    private final String message;

    private ErrorResponse(ErrorCode errorCode) {
        this.code = errorCode.name();
        this.message = errorCode.getMessage();
    }

    private ErrorResponse(ErrorCode errorCode, String message) {
        this.code = errorCode.name();
        this.message = message;
    }

    public static ResponseEntity<ErrorResponse> of(ErrorCode errorCode) {
        return ResponseEntity.status(errorCode.getStatus())
                .body(new ErrorResponse(errorCode));
    }

    public static ResponseEntity<ErrorResponse> of(ErrorCode errorCode, String message) {
        return ResponseEntity.status(errorCode.getStatus())
                .body(new ErrorResponse(errorCode, message));
    }
}

File: review-service/src/main/kotlin/com/ecommerce/reviewservice/global/exception/ErrorResponse.kt
package com.ecommerce.reviewservice.global.exception

import org.springframework.http.ResponseEntity

class ErrorResponse(val code: String, val message: String) {

    /** companion object
     * - java의 static을 대신
     * - class 하나당 최대 한개
     */
    companion object {
        fun of(errorCode: ErrorCode): ResponseEntity<ErrorResponse> {
            return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode.name, errorCode.message))
        }

        fun of(errorCode: ErrorCode, message: String): ResponseEntity<ErrorResponse> {
            return ResponseEntity.status(errorCode.status).body(ErrorResponse(errorCode.name, message))
        }
    }
}